File: problems/0494-target-sum/solution.cpp
// Link: https://leetcode.cn/problems/target-sum/
// Level: Medium
// Date: 2025-05-19

class Solution {
private:
    int backtrack(vector<int>& nums, int target, int idx, int curr_sum) {
        int n = nums.size();
        if (idx == n) {
            return curr_sum == target ? 1 : 0;
        }
        return backtrack(nums, target, idx + 1, curr_sum + nums[idx]) +
            backtrack(nums, target, idx + 1, curr_sum - nums[idx]);
    }

public:
    int findTargetSumWays(vector<int>& nums, int target) {
        int res = backtrack(nums, target, 0, 0);
        return res;
    }
};

File: problems/0494-target-sum/solution.go
// Link: https://leetcode.cn/problems/target-sum/
// Level: Medium
// Date: 2025-05-19

func findTargetSumWays(nums []int, target int) int {
	n := len(nums)
	var backtrack func(int, int)
	backtrack = func(index, currSum int) int {
		if index == n {
			if currSum == target {
				return 1
			}
			return 0
		}
		return backtrack(index+1, currSum+nums[index]) + backtrack(index+1, currSum-nums[index])
	}
	res := backtrack(0, 0)
	return res
}
